src/Simplify.elm

```diff
@@ -953,24 +1036,129 @@ errorForUnknownIgnoredConstructor list =
-        , rightSidesOfPlusPlus = []
-        , inferredConstantsDict = RangeDict.empty
+        | imports =
+            context.imports
+                |> insertImport (import_.moduleName |> Node.value)
+                    { alias =
+                        import_.moduleAlias |> Maybe.map Node.value
+                    , exposed =
+                        case import_.exposingList of
+                            Nothing ->
+                                ExposedSome Set.empty
+
+                            Just (Node _ existingExposing) ->
+                                case existingExposing of
+                                    Exposing.All _ ->
+                                        ExposedAll
+
+                                    Exposing.Explicit list ->
+                                        ExposedSome
+                                            (Set.fromList
+                                                (List.map (\(Node _ expose) -> nameOfExpose expose) list)
+                                            )
+                    }
     }
 
 
+nameOfExpose : Exposing.TopLevelExpose -> String
+nameOfExpose topLevelExpose =
+    case topLevelExpose of
+        Exposing.FunctionExpose name ->
+            name
+
+        Exposing.TypeOrAliasExpose name ->
+            name
+
+        Exposing.InfixExpose name ->
+            name
+
+        Exposing.TypeExpose { name } ->
+            name
+
+
+
+-- DECLARATION VISITOR
+
+
+declarationVisitor : Node Declaration -> ModuleContext -> ModuleContext
+declarationVisitor declarationNode context =
+    let
+        contextReset : ModuleContext
+        contextReset =
+            { context
+                | rangesToIgnore = []
+                , rightSidesOfPlusPlus = []
+                , inferredConstantsDict = RangeDict.empty
+            }
+    in
+    case Node.value declarationNode of
+        Declaration.CustomTypeDeclaration variantType ->
+            let
+                variantNames : Set String
+                variantNames =
+                    variantType.constructors
+                        |> List.map (\(Node _ variant) -> Node.value variant.name)
+                        |> Set.fromList
+
+                variantTypeName : String
+                variantTypeName =
+                    Node.value variantType.name
+            in
+            { contextReset
+                | exposedVariants =
+                    if context.exposedAll then
+                        context.exposedVariants
+                            |> Dict.insert variantTypeName
+                                variantNames
+
+                    else
+                        context.exposedVariants
+                            |> Dict.update variantTypeName
+                                (Maybe.map
+                                    (\_ ->
+                                        variantNames
+                                    )
+                                )
```

Comment:
Why blow up the amount of variant names stored for types we don't expose (I think I'm misunderstanding the question)

Reply:
Using `Dict.insert` always would not increase the number of variant names, since it will override the previous value. Also, I don't think it's in practice ever going to encounter the same `variantTypeName` twice. But maybe I'm in turn misunderstanding what this function does.

Reply:
It adds variant names for a specific type name – only if we need this information.
If we always insert the type and its variant names, even for type names that weren't added already to show they are exposed (this happens on ModuleContext init), we will collect and store information that's useless but storage space polluting

Reply:
Ah gotcha, now I see it. Alright :+1: